src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,328 @@
+
+public class ReadWriteSplittingPlugin implements IConnectionPlugin {
+  public static final int WRITER_INDEX = 0;
+  public static final String METHOD_SET_READ_ONLY = "setReadOnly";
+
+  private final ICurrentConnectionProvider currentConnectionProvider;
+  private final IConnectionProvider connectionProvider;
+  private final PropertySet propertySet;
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+
+  private boolean inTransaction = false;
+  private boolean readOnly = false;
+  private List<HostInfo> hosts = new ArrayList<>();
+  private JdbcConnection writerConnection;
+  private JdbcConnection readerConnection;
+
+  public ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this(currentConnectionProvider, new BasicConnectionProvider(), propertySet, nextPlugin, logger);
+  }
+
+  ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      IConnectionProvider connectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.connectionProvider = connectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+  }
+
+  @Override
+  public Object executeOnConnectionBoundObject(Class<?> methodInvokeOn, String methodName, Callable<?> executeSqlFunc, Object[] args) throws Exception {
+    return this.nextPlugin.executeOnConnectionBoundObject(methodInvokeOn, methodName, executeSqlFunc, args);
+  }
+
+  @Override
+  public Object executeOnConnection(Method method, List<Object> args) throws Exception {
+    if (METHOD_SET_READ_ONLY.equals(method.getName()) && args != null && args.size() > 0) {
+      switchConnectionIfRequired((Boolean) args.get(0));
+    }
+    return this.nextPlugin.executeOnConnection(method, args);
+  }
+
+  void switchConnectionIfRequired(Boolean readOnly) throws SQLException {
+    if (readOnly == null) return;
+    final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    if (readOnly) {
+      if (!this.inTransaction && (!isReaderConnection() || currentConnection.isClosed())) {
+        try {
+          switchToReaderConnection(currentConnection);
+        } catch (SQLException e) {
+          if (!isConnectionUsable(currentConnection)) {
+            throw new SQLException(
+                Messages.getString("ReadWriteSplittingPlugin.1"),
+                MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE, e);
+          }
+
+          // Failed to switch to a reader; use current connection as a fallback
+          changeReaderConnectionTo(currentConnection);
+        }
+      }
+    } else {
+      if (!isWriterConnection() || currentConnection.isClosed()) {
+        if(this.inTransaction) {
```

Comment:
```suggestion
        if (this.inTransaction) {
```

Reply:
I am a little bit confused by this check.
If `!isWriterConnection()` returned true, so we are currently on a reader connection, is it possible for `inTransaction` to be true?

Reply:
I believe you can start transactions on reader connections, for example "START TRANSACTION READ ONLY" is valid in MySQL (see [here](https://dev.mysql.com/doc/refman/5.6/en/commit.html))